### DISCUSSION TITLE
How to name isLeft and the sides for nodes with 90° rotated layout?

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#720] [General]
Hello,

in the new "horizontal" layout (where child nodes have the same horizontal position) "left" becomes "above" and "right" becomes "below". But the layout can be changed to back to top-bottom at any time, and probably even set for individual nodes. So I need a more general term that could fit the both directions. And I don't have any idea how to call it.

Do you have any?

Cheers, Dimitry

**Comment by euu2021** [comment]
I can't think of any word that is represents both the concepts of left/above or right/below. Maybe, using a generic Side A and Side B?

**Comment by macmarrum** [comment]
I propose to use compass directions:
- northwest (NW)
- southeast (SE)

<img src="https://st4.depositphotos.com/1915807/20350/v/950/depositphotos_203503180-stock-illustration-compass-direction-north-west-east.jpg" width="300">

**Comment by dpolivaev** [comment]
Because the left and right sides are frequently used, I am thinking about introducing a new layout property

~~~~~
public enum LayoutOrientation {
    NORTH, EAST;
}
~~~~~
so that child node side values "left" and "right" can be interpreted depending on its value.

In the UI, the user could choose between "vertical" and "horizontal" layouts.

**Reply by dpolivaev** [reply]
@macmarrum @euu2021 please comment

**Reply by dpolivaev** [reply]
The enum could also be called `LayoutDirection`

**Reply by macmarrum** [reply]
I've had a look at https://dictionary.cambridge.org/dictionary/english/orientation

> orientation noun [U] (ARRANGEMENT)
C1 formal
[arrangement](https://dictionary.cambridge.org/dictionary/english/arrangement) or [direction](https://dictionary.cambridge.org/dictionary/english/direction):
_The [building](https://dictionary.cambridge.org/dictionary/english/building) has an east-west orientation (= it is [built](https://dictionary.cambridge.org/dictionary/english/built) on a [line](https://dictionary.cambridge.org/dictionary/english/line) between [east](https://dictionary.cambridge.org/dictionary/english/east) and [west](https://dictionary.cambridge.org/dictionary/english/west))._

Assuming my proposal to use NW and SE instead of LEFT/UP and RIGHT/DOWN has been accepted and taking the above into consideration, I propose
- **Left**-right layout -> **WEST**_EAST
- **Up**-down layout -> **NORTH**_SOUTH

```java
public enum LayoutOrientation {
    WEST_EAST, NORTH_SOUTH;
}
```

**Reply by dpolivaev** [reply]
Actually, I have something different in mind.
LayoutOrientation NORTH means that left and right are as on your display.
LayoutOrientation EAST means that the left is at the top and the right is at the bottom.

**Reply by euu2021** [reply]
>LayoutOrientation EAST means that the left is at the top and the right is at the bottom.

Do you mean WEST? Like: `LayoutOrientation WEST means that the left is at the top and the right is at the bottom.`

**Reply by dpolivaev** [reply]
North is the top.
If I look to the East, the North is to my left. Isn't it?

**Reply by euu2021** [reply]
Maybe I misunderstood, but, if left becomes top, then what happens is a clockwise rotation, so the layout orientation is to the West, ins't it:

![image](https://user-images.githubusercontent.com/77707706/194716686-400de09b-ab8f-48c9-a365-a24931e5c002.png)

**Reply by dpolivaev** [reply]
I do not rotate the north, the north is always at the top of the monitor.
I rotate the observer originally looking to the north so that they look to the east. And then the top of the display AKA north is to their left.

**Reply by macmarrum** [reply]
> I do not rotate the north, the north is always at the top of the monitor.
I rotate the observer originally looking to the north so that they look to the east. And then the top of the display AKA north is to their left.

This logic is hard for me to process. It feels unnatural, because it's not the observer who is rotated. It's the map, isn't it?

A left-right map, where left points to the West, is rotated 90* so that the "left" becomes "up" and points to the North, and the nodes grow up and down.

I like the building example: _The [building](https://dictionary.cambridge.org/dictionary/english/building) has an east-west orientation (= it is [built](https://dictionary.cambridge.org/dictionary/english/built) on a [line](https://dictionary.cambridge.org/dictionary/english/line) between [east](https://dictionary.cambridge.org/dictionary/english/east) and [west](https://dictionary.cambridge.org/dictionary/english/west))._

Using this logic, a mindmap has an east-west orientation (nodes grow left-right between east and west) or a north-south one.

**Reply by euu2021** [reply]
west-east/north-south feels the most intuitive to me, indeed

**Reply by dpolivaev** [reply]
West-east and north-south directions are obvious, but they don't give us any interpretation for left an right nodes. Only my proposal solves it.

**Reply by ldmpub** [reply]
What about?

* Is forward / clockwise
* Is backward / anti-clockwise

According to me it's working both for Horizontal West-east layout and (new) Vertical layout North-south ...

Above and Below can still be used whatever orientation referring to "node childhood display level".

**Reply by macmarrum** [reply]
> West-east and north-south directions are obvious, but they don't give us any interpretation for left an right nodes. Only my proposal solves it.

It might be technically correct, yet quite unnatural and difficult to understand and explain to people.

I propose to use `enum LayoutOrientation { WEST_EAST, NORTH_SOUTH; }` and `enum Side { NORTHWEST, SOUTHEAST, DEFAULT }` and `Node.isNorthWest()` in place of `Node.isLeft()`

Alternatively `enum Side { NORTH_OR_WEST, SOUTH_OR_EAST, DEFAULT }`

If you'd rather retain `enum Side { LEFT, RIGHT, DEFAULT }` and `Node.isLeft()`, then I'd propose to use `enum LayoutOrientation { REGULAR, ROTATED; }`. With this approach, `isLeft()` for `ROTATED` will have the meaning of `isUp()` and `RIGHT` will mean `DOWN`. But I like the first approach much better.

**Reply by dpolivaev** [reply]
I am still thinking. 
And I want you to know that I am going to make LayoutOrientation a setting that can be set for each node individually.
Then it should be possible to choose if the child nodes are placed on one particular side or at both sides for non-root nodes.

**Reply by macmarrum** [reply]
That's great :heart:

**Reply by dpolivaev** [reply]
What do you think about `enum LayoutOrientation { LEFT_TO RIGHT, TOP_DOWN; }` and `enum Side { TOP_LEFT, BOTTOM_RIGHT, DEFAULT }` and Node.isTopLeft() in place of `Node.isLeft()`?

`LEFT` and `isLeft()` should be kept as deprecated so that old scripts still can work.

**Reply by macmarrum** [reply]
I like the idea :+1: 
I think TOP_BOTTOM would be better than TOP_DOWN. Otherwise it looks great
